File: src/ns/rack-device-view.ts
import { Ableton } from "../index.js";
import { Namespace } from "./index.js";
import { Chain, RawChain } from "./chain.js";
import { DrumPad, RawDrumPad } from "./drum-pad.js";

export interface GettableProperties {
  /** Drum racks only; `null` on other rack types. */
  drum_pads_scroll_position: number | null;
  is_collapsed: boolean;
  is_showing_chain_devices: boolean;
  selected_chain: RawChain | null;
  /** Drum racks only; `null` on other rack types. */
  selected_drum_pad: RawDrumPad | null;
}

export interface TransformedProperties {
  selected_chain: Chain | null;
  selected_drum_pad: DrumPad | null;
}

export interface SettableProperties {
  drum_pads_scroll_position: number;
  is_collapsed: boolean;
  is_showing_chain_devices: boolean;
  /** Live object id of the chain to select, or `null` to clear. */
  selected_chain: string | null;
  /** Live object id of the drum pad to select. */
  selected_drum_pad: string;
}

export interface ObservableProperties {
  drum_pads_scroll_position: number | null;
  is_collapsed: boolean;
  is_showing_chain_devices: boolean;
  selected_chain: RawChain | null;
  selected_drum_pad: RawDrumPad | null;
}

/**
 * View aspects of a rack device.
 */
export class RackDeviceView extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(ableton: Ableton, nsid: string) {
    super(ableton, "rack-device-view", nsid);

    this.transformers = {
      selected_chain: (chain) => (chain ? new Chain(ableton, chain) : null),
      selected_drum_pad: (pad) => (pad ? new DrumPad(ableton, pad) : null),
    };

    this.cachedProps = {
      selected_chain: true,
      selected_drum_pad: true,
    };
  }
}

File: midi-script/Namespaces/RackDeviceView.py

from .Chain import Chain
from .DrumPad import DrumPad
from .Interface import Interface


class RackDeviceView(Interface):
    def get_ns(self, nsid):
        return Interface.get_obj(nsid).view

    def get_selected_chain(self, ns):
        return Chain.serialize_chain(ns.selected_chain)

    def set_selected_chain(self, ns, value):
        ns.selected_chain = Interface.get_obj(value) if value else None

    def get_selected_drum_pad(self, ns):
        try:
            return DrumPad.serialize_drum_pad(ns.selected_drum_pad)
        except RuntimeError:
            return None

    def set_selected_drum_pad(self, ns, value):
        ns.selected_drum_pad = Interface.get_obj(value)

    def get_drum_pads_scroll_position(self, ns):
        try:
            return ns.drum_pads_scroll_position
        except RuntimeError:
            return None

    def set_drum_pads_scroll_position(self, ns, value):
        ns.drum_pads_scroll_position = int(value)

    def set_is_showing_chain_devices(self, ns, value):
        ns.is_showing_chain_devices = bool(value)
